internal/app/machined/pkg/system/services/registry/registry.go

```diff
@@ -0,0 +1,248 @@
+	if err != nil {
+		return xerrors.NewTaggedf[internalErrorTag]("failed to get content reader: %w", err)
+	}
+
+	defer reader.Close() //nolint:errcheck
+
+	http.ServeContent(w, req, ref.Digest().String(), info.UpdatedAt, &readSeeker{ReaderAt: reader, Size: info.Size})
+
+	return nil
+}
+
+func (svc *Service) resolveCanonicalRef(p params) (reference.Canonical, error) {
+	ref, err := reference.ParseDockerRef(p.String())
+	if err != nil {
+		return nil, xerrors.NewTaggedf[badRequestTag]("failed to parse docker ref: %w", err)
+	}
+
+	cRef, ok := ref.(reference.Canonical)
+	if ok {
+		return cRef, nil
+	}
+
+	namedTagged, ok := ref.(reference.NamedTagged)
+	if !ok {
+		return nil, xerrors.NewTaggedf[internalErrorTag]("incorrect reference type: %T", ref)
+	}
+
+	taggedFile := filepath.Join(svc.root, "manifests", namedTagged.Name(), "reference", namedTagged.Tag())
+
+	ntSum, err := hashFile(taggedFile)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return nil, xerrors.NewTaggedf[internalErrorTag]("failed to hash manifest: %w", err)
+		}
+
+		return nil, xerrors.NewTagged[notFoundTag](err)
+	}
+
+	sha256file := filepath.Join(svc.root, "manifests", namedTagged.Name(), "digest", digest.NewDigestFromBytes(digest.SHA256, ntSum).String())
+
+	sSum, err := hashFile(sha256file)
+	if err != nil {
+		return nil, xerrors.NewTaggedf[internalErrorTag]("failed to hash '%x': %w", sSum, err)
+	}
+
+	if !bytes.Equal(ntSum, sSum) {
+		return nil, xerrors.NewTaggedf[internalErrorTag]("hash for '%s' is not equal for hash to '%s'", taggedFile, sha256file)
+	}
+
+	return &canonical{
+		NamedTagged: namedTagged,
+		digest:      digest.NewDigestFromBytes(digest.SHA256, ntSum),
+	}, nil
+}
+
+func hashFile(f string) (_ []byte, returnErr error) {
+	data, err := os.Open(f)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() { returnErr = cmp.Or(returnErr, data.Close()) }()
+
+	h := sha256.New()
+	if _, err = io.Copy(h, data); err != nil {
+		return nil, err
+	}
+
+	return h.Sum(nil), nil
+}
+
+func getManifestData(ctx context.Context, store content.Store, ref reference.Canonical) (string, []byte, error) {
+	manifestBlob, err := content.ReadBlob(ctx, store, ocispec.Descriptor{Digest: ref.Digest()})
+	if err != nil {
+		return "", nil, xerrors.NewTaggedf[internalErrorTag]("failed to read content blob: %w", err)
+	}
+
+	var manifest struct {
+		MediaType string `json:"mediaType"`
+	}
+
+	if err = json.Unmarshal(manifestBlob, &manifest); err != nil {
+		return "", nil, xerrors.NewTaggedf[internalErrorTag]("failed to unmarshal manifest: %w", err)
+	}
+
+	if manifest.MediaType == "" {
+		return "", nil, xerrors.NewTaggedf[internalErrorTag]("media type is empty")
+	}
+
+	return manifest.MediaType, manifestBlob, nil
+}
+
+type canonical struct {
+	reference.NamedTagged
+	digest digest.Digest
+}
+
+func (c *canonical) String() string        { return c.NamedTagged.String() + "@" + c.digest.Encoded() }
+func (c *canonical) Digest() digest.Digest { return c.digest }
```

Comment:
Weird formatting, won't classic one with newlines look less out of place?

Reply:
Usually gofmt allows functions with 105 characters or less to be single line.